src/syntax/instrumentPpx.ml

```diff
@@ -354,26 +340,43 @@ class instrumenter = object (self)
           l
         in
           Str.value ~loc rec_flag l
-    | Pstr_eval (e, a) ->
+    | Pstr_eval (e, a) when not (attribute_guard || extension_guard) ->
         Str.eval ~loc (wrap_expr Common.Toplevel_expr (self#expr e))
     | _ ->
         super#structure_item si
 
-  (*  Initializes storage and applies requested marks. *)
+  (* Guard these because they can carry payloads that we
+     do not want to instrument. *)
+  method! extension e =
+    extension_guard <- true;
+    let r = super#extension e in
+    extension_guard <- false;
+    r
+
+  method! attribute a =
+    attribute_guard <- true;
+    let r = super#attribute a in
+    attribute_guard <- false;
+    r
+
+  (* Initializes storage and applies requested marks. *)
   method! structure ast =
-    let ast = super#structure ast in
-    match get_filename ast with
-    | None -> ast
-    | Some file ->
-      if not (InstrumentState.is_file file) then
-        let header =
-          match !InstrumentArgs.mode with
-          | InstrumentArgs.Safe   -> safe file
-          | InstrumentArgs.Fast
-          | InstrumentArgs.Faster -> faster file
-        in
-        header :: ast
-      else
-        ast
+    (*let ts = String.concat "," (List.map typoo ast) in *)
+    if extension_guard || attribute_guard then
```

Comment:
@rleonid, I am currently reviewing some Bisect_ppx code. Does this test do anything in this function? What was the intent?

This is what I think happens: `extension_guard` or `attribute_guard` triggers `super#structure ast` on the next line (obviously). However, I believe it's redundant.

If `extension_guard` or `attribute_guard` is `true`, that means

1. We recursively traversed into the AST, so `get_filename ast` (or its alternative in the modern version of this code) did not evaluate to `None`.
2. The file has already been added to `InstrumentState`, as we are in a nested structure, and must have seen the file's top-level structure before.

That means that if we deleted this guard, the code in the `else` case would still fall through to `super#structure ast`, on line 380 at this point in history. The modern code does largely the same thing, just rearranged a bit.

Any insight?

Perhaps also @smondet on what the issue here was.